### PR TITLE
Ethereum types refactor

### PR DESCRIPTION
**Pull request by debris**
@NikVolf @rphmeier @tomusdrw tell me what you think. to see the interface do `cargo doc -p ethereum-types`

### CODE REVIEW

**Comment by NikVolf** [comment]
~~wow, but we had been using `bigint` from crates.io for quite a time already, no?~~

disregard this :)

**Comment by NikVolf** [comment]
@debris 
we already have `parity-hash` crate with dependency on `bigint`, why not just use that?

File: ethereum-types/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#![cfg_attr(not(feature="std"), no_std)]
```
Comment (tomusdrw) [inline ethereum-types/src/lib.rs:1]: Would be good to warn on missing docs.

File: ethereum-types/src/uint.rs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+	/// Multiplies two 256-bit integers to produce full 512-bit integer
+	/// No overflow possible
+	#[inline(always)]
+	#[cfg(not(all(asm_available, target_arch="x86_64")))]
+	pub fn full_mul(self, other: U256) -> U512 {
+		U512(uint_full_mul_reg!(U256, 4, self, other))
+	}
+}
+
+impl From<U256> for U512 {
```
Comment (tomusdrw) [inline ethereum-types/src/uint.rs:154]: Can't this be turned into macro in `uint` crate? Like:
```
impl_upcast_from!(U256,4 for U512,8)
// Although this doesn't seem right (`TryFrom` would be better)
impl_downcast_from!(U512,8, for U256,4)
```

File: ethereum-types/src/uint.rs [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+		let mut ret = [0; 8];
+		ret[0] = arr[0];
+		ret[1] = arr[1];
+		ret[2] = arr[2];
+		ret[3] = arr[3];
+		U512(ret)
+	}
+}
+
+impl From<U512> for U256 {
```
Comment (tomusdrw) [inline ethereum-types/src/uint.rs:166]: Do we really need panicking `From` implementation?

File: ethereum-types/src/uint.rs [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+	}
+}
+
+impl From<U256> for u32 {
+	fn from(value: U256) -> u32 {
+		value.as_u32()
+	}
+}
+
+impl<'a> From<&'a [u8; 32]> for U256 {
```
Comment (tomusdrw) [inline ethereum-types/src/uint.rs:266]: Why isn't this part of the `impl_uint!`?

**Review by tomusdrw — APPROVED** [review]
Looks good to me (modulo small grumbles). I like how now `uint` is a create that allows you to implement and name any type and doesn't come with pre-defined implementations that we happen to use in ethereum.

I think we might also considering adding serializable ethereum-types as a separate crate or at least serialization implementations so that you can use:
```rust
#[derive(Serialize)]
struct MyType {
 #[serde(serialize_with="ethereum_types_serialization::address")]
 pub address: ethereum_types::Address,
}
```

**Comment by NikVolf** [comment]
Hash really should be in separate crate

This crate is just for integer arithmetic and stable for a long time

**Comment by tomusdrw** [comment]
It is in a separate crate:
`bigint` - `impl_uint` macro
`fixed-hash` - `impl_hash` macro
`ethereum-types` - Ethereum-specific instances of those types (both H256 & U256)

**Comment by NikVolf** [comment]
Yeah, but why in this repository?

File: uint/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+[package]
+description = "Large fixed-size integers arithmetics"
+homepage = "http://parity.io"
+repository = "https://github.com/ethcore/bigint"
```
Comment (NikVolf) [inline uint/Cargo.toml:4]: what is `ethcore/bigint` ?

**Comment by debris** [comment]
@NikVolf the idea for this pr is to separate uint/hash crates from their usage. Mostly to clean existing architecture where different crates where using different hashes and where it is weird that some types are implemented by default why the others are not

eg. `H256`, `H1025`, `U256` are implemented but, it's not clearly described why `U1024` is not.

Also, I believe that reusing hash and integer type is now easier for polkadot and other projects

I made a pr in this repo, cause I wanted to put it up for a discussion. I agree, maybe we should have a new repo for it? or rename this one?

**Comment by NikVolf** [comment]
@debris
I totally agree with your motivation,
code looks great, i just think `bigint` should be left alone as a stable library in this dedicated repository

**Pull request closed without merging**